### ISSUE TITLE
Getting a return value from runsource?

### ISSUE THREAD

**Issue by dsblank** [rfk/pypyjs#63]
In @minrk 's new [PyPy.js kernel](https://github.com/minrk/pypyjskernel) for Jupyter, he has:

```javascript
        code = "r = c.runsource('''" + code + "''', '<input>', 'exec')";
        vm.eval(code).then(function() {
            _finish_execute.apply(that, [r, request, true]);
        });
```

but that doesn't make expressions like `1 + 1` give 2 as a return value. In fact, I don't think anything is ever returned. I know that many Python implementations have alternate ways of interpreting code, that include the things that one types in a REPL. Is there an alternative to 'exec' for this here? Or another method to get a return value?

**Comment by rfk** [comment]
I'm not sure I fully understand the question, but here's some high-level info that may help.

The PyPy.js `vm.eval()` function executes chunks of code as statements, and hence does not produce a return value besides success/failure.  It's more akin to the python `exec` statement than the python `eval` function, and perhaps we should consider re-naming it to avoid confusion.

Currently, the only way to get a return value out the the PyPy.js vm is to use its `get` method to read values out of the interpreter global namespace.  So for example you can do:

```
vm.eval("r = 1 + 1").then(function() {
  return vm.get("r"), function(r) {
    console.log(r); // should print "2"
  }
})
```

**Comment by dsblank** [comment]
Thanks, that answered the question! I have done that hack before with different Python implementations. 

The issue is to try to make a REPL that can take an expression or statement(s). IronPython has `Microsoft.Scripting.SourceCodeKind.InteractiveCode` for use in their `engine.CreateScriptSourceFromString`. I think IPython parses and looks at the AST to decide to eval or exec (and can even return the last expr in a list of statements).

Maybe a workaround is to try to compile as a expr, and if fail, then fallback to compile as a statement? That would at least get the simple case of a single expression separated, yes?

**Comment by rfk** [comment]
> Maybe a workaround is to try to compile as a expr, and if fail, then fallback to compile as a statement? 

This seems sane.  Do you think it makes sense to provide helpers like this as an API on PyPy.js, or is it just code that needs to grow in the jupyter kernel?  (And if the former, I'll happily take suggestions for what such an API should look like)

**Comment by dsblank** [comment]
If that is code that can be written effectively here, I think that makes sense. Anyone that wants to make a REPL out of PyPy.js would possibly be interested in such an interface. 

Perhaps:

```python
runsource(CODE, SOURCE, MODE)
```
where MODE is one of:
* `eval`
* `exec`
* `interactive`, or `repl`, or ...?

**Comment by perkinslr** [comment]
Makes sense to ship it with something, otherwise everyone implementing a 'execute python code from js' (repl or otherwise) will end up having to implement their own.  I don't think it makes sense to have code fragmentation happen happen on such a fundamental thing, better to avoid the 'more than one way to do it' host of issues.

**Comment by rfk** [comment]
In the latest version I've split this into separate `eval()` and `exec()` methods that mirror those from python.  `eval()` works on expressions and returns a value, `exec()` takes whole statement suites and does not return anything.  There's also a much higher-level `repl()` method that encompasses all the details of the read-eval-print loop.

I don't think this *quite* addresses what you were after here, since it doesn't let you distinguish between expressions or statements just by looking at the code.  But I think it's a step in the right direction.
